packages/eslint-plugin-choreography-ts/src/no-unused-colocally-location.ts

```diff
@@ -0,0 +1,166 @@
+// no-unused-colocally-location rule
+// All locations specified for a `colocally` call must be used inside the body
+// https://github.com/shumbo/choreography-ts/issues/7
+
+"use strict";
+import { AST_NODE_TYPES, TSESLint, TSESTree } from "@typescript-eslint/utils";
+
+type MessageIDs = "warning";
+
+const choreographySelector = `VariableDeclaration[kind = "const"] > VariableDeclarator[id.typeAnnotation.typeAnnotation.typeName.name = "Choreography"]`;
+const colocallySelector = `${choreographySelector} CallExpression[callee.name = "colocally"]`;
+const colocallyColocallyUseLocater = `${colocallySelector} CallExpression[callee.name = "colocally"]`;
+const colocallyLocallyUseLocater = `${colocallySelector} CallExpression[callee.name = "locally"]`;
+const colocallyMulticastUseLocater = `${colocallySelector} CallExpression[callee.name = "multicast"]`;
+const colocallyCommUseLocater = `${colocallySelector} CallExpression[callee.name = "comm"]`;
+const colocallyBroadcastUseLocater = `${colocallySelector} CallExpression[callee.name = "broadcast"]`;
+const colocallyTypeUseLocater = `${colocallySelector} TSLiteralType`;
+const choreographyAtExit = `Program > ${choreographySelector}:exit`;
+
+// Implementation: Check whether each location passed to `colocally` is present inside the body of the choreography argument
+// as an argument to any operator that accepts location parameters (locally, colocally, comm, multicast, broadcast)
+// or as a type argument
+const noUnusedColocallyLocation: TSESLint.RuleModule<MessageIDs, []> = {
+  defaultOptions: [],
+  meta: {
+    type: "problem",
+    docs: {
+      description:
+        "All locations specified for `colocally` must be used inside the body of the call",
+      recommended: "recommended",
+      url: "https://github.com/shumbo/choreography-ts/issues/7",
+    },
+    fixable: undefined, // not automatically fixable
+    hasSuggestions: false, // no suggested fixes available
+    messages: {
+      warning: "Location `{{ location }}` is not used inside the body.",
+    },
+    schema: [],
+  },
+  create(context) {
+    // Store locations for each `colocally` usage
+    let Locations: {
```

Comment:
why not store data as `Map<string, TSESLint.ReportDescriptor<MessageIDs>`? 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

Then we can use `set` instead of `push` and `delete` instead of `filter`.

Reply:
The only problem with this is if we have the same location being used across multiple nested colocally calls, in which case we might want to have multiple instances of the same location but from different parts of the code. I don't think a map would work here for this reason.  We really want an array of objects to work with here where we filter out the ones that match the current location node being looked at in the AST. I don't think a `Map` would help here, and certainly not `Map.delete`.

Reply:
@ImpregnableProgrammer I don't quite understand your point here. I understand that the code might have multiple instances of the same location, but `filter` removes all elements that do not satisfy the predicate, so I don't see how  `delete` would work differently.

Reply:
> @ImpregnableProgrammer I don't quite understand your point here. I understand that the code might have multiple instances of the same location, but `filter` removes all elements that do not satisfy the predicate, so I don't see how `delete` would work differently.

Ok, I think I see what you mean. Since we are replacing that location once we encounter a use of it in a `colocally` expression, a `Map` might actually work then between the location and a report for it. I’ll try to make the change then.

Reply:
I actually realized that I'll still have to use an array of `Map<string, TSESLint.ReportDescriptor<MessageIDs>` since I still need to preserve the contextual use of locations. Otherwise with a single `Map`, a location from an outer `colocally` context that is used after a nested `colocally` call would still be marked as unused once the inner `colocally` node is traversed. This then does not differ much from the previous implementation, except perhaps being slightly faster.

Reply:
Ok, this is done. Feel free to re-review when you can!